docs/migrations/README.md

```diff
@@ -0,0 +1,339 @@
+
+*Decoding:*
+When decoding, TxBody.UnpackInterfaces will need a special case to detect if Any type URLs match the service method format (ex. /cosmos.gov.Msg/SubmitProposal) by checking for two / characters. Messages that are method names plus request parameters instead of a normal Any messages will get unpacked into the ServiceMsg struct:
+
+type ServiceMsg struct {
+  // MethodName is the fully-qualified service name
+  MethodName string
+  // Request is the request payload
+  Request MsgRequest
+}
+
+#### Keeper
+The `Keeper` constructor now takes a `codec.Marshaler` instead of a concrete Amino codec. This exact type provided is 
+specified by `ModuleCdc`.
+
+#### module.go
+* `type AppModuleBasic struct{}` is  updated to:
+
+    ```go
+    type AppModuleBasic struct {
+        cdc codec.Marshaler
+    }
+    ```
+* `RegisterCodec(cdc *codec.Codec)` method is changed to `RegisterLegacyAminoCodec(cdc *codec.LegacyAmino)`
+* Added `RegisterInterfaces` method which implements `AppModuleBasic` which takes one parameter of type 
+`"github.com/cosmos/cosmos-sdk/codec/types".InterfaceRegistry`. This method is used for registering interface types of module.
+    ```go
+    func (AppModuleBasic) RegisterInterfaces(registry cdctypes.InterfaceRegistry) {
+        // register all interfaces in module.
+	    types.RegisterInterfaces(registry) //module's types/codec.go
+    }
+    ````
+* `DefaultGenesis()` takes codec input now
+    ```go
+    func (AppModuleBasic) DefaultGenesis(cdc codec.JSONMarshaler) json.RawMessage {}
+    ```
+* `ValidateGenesis` now requires `Marshaler`, `TxEncodingConfig`, `json.RawMessage` as input.
+    ```go
+    func (AppModuleBasic) ValidateGenesis(cdc codec.JSONMarshaler, config client.TxEncodingConfig, bz json.RawMessage) error {}
+    ```
+* `GetQueryCmd(cdc *codec.Codec)`,`GetTxCmd(cdc *codec.Codec)` is changed to `GetQueryCmd()`,`GetTxCmd()` respectively.
+* Return type of `Route()` method which implements `AppModule` is changed from `string` to `"github.com/cosmos/cosmos-sdk/types".Route`. We will return a NewRoute which includes `RouterKey` and `NewHandler` as params.
+    ```go
+    func (am AppModule) Route() sdk.Route {
+        return sdk.NewRoute(types.RouterKey, handler.NewHandler(am.keeper))
+    }
+    ```
+* `QuerierHandler` is renamed to `LegacyQuerierHandler`.
+  ```go
+    func (am AppModule) LegacyQuerierHandler(legacyQuerierCdc *codec.LegacyAmino) sdk.Querier {}
+  ```
+
+* `RegisterServices` is added newly and it registers module's `MsgServer` and gRPC's `QueryServer`.
+  ```go
+  func (am AppModule) RegisterServices(cfg module.Configurator) {
+      types.RegisterMsgServer(cfg.MsgServer(), keeper.NewMsgServerImpl(am.keeper))
+      types.RegisterQueryServer(cfg.QueryServer(), am.keeper)
+  }
+  ```
+* `RegisterGRPCGatewayRoutes` is added newly which is for registering module's gRPC gateway routes with API Server.`RegisterQueryHandlerClient` is auto generated by proto code generator.
```

Comment:
these items look like CHANGELOG entries, I think it's a bit useless to put them here in this format. I can think about this.